File: mobile/lib/features/issues/domain/issue.dart
import '../../../core/models/department.dart';
import '../../../core/models/employee.dart';
import '../../../core/models/organization.dart';

class IssueActivity {
  IssueActivity({required this.id, required this.action, this.description, this.causerName, required this.createdAt});

  final int id;
  final String action;
  final String? description;
  final String? causerName;
  final String createdAt;

  factory IssueActivity.fromJson(Map<String, dynamic> json) => IssueActivity(
        id: json['id'] as int,
        action: json['action'] as String? ?? '',
        description: json['description'] as String?,
        causerName: json['causer_name'] as String?,
        createdAt: json['created_at'] as String? ?? '',
      );
}

class IssueComment {
  IssueComment({required this.id, this.userName, required this.body, required this.createdAt});

  final int id;
  final String? userName;
  final String body;
  final String createdAt;

  factory IssueComment.fromJson(Map<String, dynamic> json) => IssueComment(
        id: json['id'] as int,
        userName: json['user_name'] as String?,
        body: json['body'] as String? ?? '',
        createdAt: json['created_at'] as String? ?? '',
      );
}

class IssueCategory {
  IssueCategory({required this.id, required this.name, required this.code});

  final int id;
  final String name;
  final String code;

  factory IssueCategory.fromJson(Map<String, dynamic> json) => IssueCategory(
        id: json['id'] as int,
        name: json['name'] as String? ?? '',
        code: json['code'] as String? ?? '',
      );
}

class IssueOrganizationOption {
  IssueOrganizationOption({required this.id, required this.name});

  final int id;
  final String name;

  factory IssueOrganizationOption.fromJson(Map<String, dynamic> json) =>
      IssueOrganizationOption(id: json['id'] as int, name: json['name'] as String? ?? '');
}

/// What the report form offers this user — decided by the server so the
/// app never re-derives role rules: which organizations they may file
/// against, the categories, and whether they must name a responsible
/// employee (a department-manager is responsible for what they report).
class IssueOptions {
  IssueOptions({required this.organizations, required this.categories, required this.mustChooseResponsible});

  final List<IssueOrganizationOption> organizations;
  final List<IssueCategory> categories;
  final bool mustChooseResponsible;

  factory IssueOptions.fromJson(Map<String, dynamic> json) => IssueOptions(
        organizations: (json['organizations'] as List<dynamic>? ?? [])
            .map((e) => IssueOrganizationOption.fromJson(e as Map<String, dynamic>))
            .toList(),
        categories: (json['categories'] as List<dynamic>? ?? [])
            .map((e) => IssueCategory.fromJson(e as Map<String, dynamic>))
            .toList(),
        mustChooseResponsible: json['must_choose_responsible'] as bool? ?? false,
      );
}

class IssueResponsibleCandidate {
  IssueResponsibleCandidate({required this.id, required this.fullName, this.department, this.position});

  final int id;
  final String fullName;
  final String? department;
  final String? position;

  String get subtitle => [position, department].whereType<String>().join(' · ');

  factory IssueResponsibleCandidate.fromJson(Map<String, dynamic> json) => IssueResponsibleCandidate(
        id: json['id'] as int,
        fullName: json['full_name'] as String? ?? '',
        department: json['department'] as String?,
        position: json['position'] as String?,
      );
}

class Issue {
  Issue({
    required this.id,
    this.reporter,
    this.organization,
    this.department,
    this.category,
    this.responsible,
    required this.title,
    this.description,
    this.objectName,
    required this.latitude,
    required this.longitude,
    required this.status,
    this.resolutionNote,
    this.resolvedByName,
    this.resolvedAt,
    required this.comments,
    required this.activities,
    required this.createdAt,
  });

  final int id;
  final Employee? reporter;
  final Organization? organization;
  final Department? department;
  final IssueCategory? category;
  final Employee? responsible;
  final String title;
  final String? description;
  final String? objectName;
  final double latitude;
  final double longitude;
  final String status;
  final String? resolutionNote;
  final String? resolvedByName;
  final String? resolvedAt;
  final List<IssueComment> comments;
  final List<IssueActivity> activities;
  final String createdAt;

  bool get isOpen => status == 'open';

  factory Issue.fromJson(Map<String, dynamic> json) => Issue(
        id: json['id'] as int,
        reporter: json['reporter'] is Map<String, dynamic>
            ? Employee.fromJson(json['reporter'] as Map<String, dynamic>)
            : null,
        organization: json['organization'] is Map<String, dynamic>
            ? Organization.fromJson(json['organization'] as Map<String, dynamic>)
            : null,
        department: json['department'] is Map<String, dynamic>
            ? Department.fromJson(json['department'] as Map<String, dynamic>)
            : null,
        category: json['category'] is Map<String, dynamic>
            ? IssueCategory.fromJson(json['category'] as Map<String, dynamic>)
            : null,
        responsible: json['responsible'] is Map<String, dynamic>
            ? Employee.fromJson(json['responsible'] as Map<String, dynamic>)
            : null,
        title: json['title'] as String? ?? '',
        description: json['description'] as String?,
        objectName: json['object_name'] as String?,
        latitude: (json['latitude'] as num).toDouble(),
        longitude: (json['longitude'] as num).toDouble(),
        status: json['status'] as String? ?? 'open',
        resolutionNote: json['resolution_note'] as String?,
        resolvedByName: json['resolved_by_name'] as String?,
        resolvedAt: json['resolved_at'] as String?,
        comments: (json['comments'] as List<dynamic>? ?? [])
            .map((e) => IssueComment.fromJson(e as Map<String, dynamic>))
            .toList(),
        activities: (json['activities'] as List<dynamic>? ?? [])
            .map((e) => IssueActivity.fromJson(e as Map<String, dynamic>))
            .toList(),
        createdAt: json['created_at'] as String? ?? '',
      );
}

File: backend/app/Models/Issue.php
<?php

namespace App\Models;

use App\Enums\IssueStatus;
use Database\Factories\IssueFactory;
use Illuminate\Database\Eloquent\Attributes\Fillable;
use Illuminate\Database\Eloquent\Builder;
use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;
use Illuminate\Database\Eloquent\Relations\BelongsToMany;
use Illuminate\Database\Eloquent\Relations\HasMany;

#[Fillable([
    'reporter_employee_id', 'organization_id', 'department_id',
    'issue_category_id',
    'title', 'description', 'object_name', 'latitude', 'longitude',
    'status', 'resolution_note', 'resolved_by', 'resolved_at',
])]
class Issue extends Model
{
    /** @use HasFactory<IssueFactory> */
    use HasFactory;

    /**
     * Central leadership sees every issue — deliberately not the broader
     * `hasCentralAccess()` set (which also includes hr and safety-manager).
     */
    public static function seesAllIssues(User $user): bool
    {
        return $user->hasAnyRole(['technical-policy', 'central-admin', 'super-admin']);
    }

    /**
     * Issues this user may see in the list, on the map and in the report:
     * what they reported or are executing, plus their whole organization for
     * an organization-admin and what their department raised or is executing
     * for a department-manager. `IssuePolicy::view()` mirrors this rule.
     */
    public function scopeVisibleTo(Builder $query, User $user): void
    {
        if (self::seesAllIssues($user)) {
            return;
        }

        $employee = $user->employee;

        // Qualified column names: the report joins organizations/categories.
        $query->where(function ($scope) use ($user, $employee) {
            $scope->where('issues.reporter_employee_id', $employee?->id ?? 0)
                ->orWhereHas('executors', fn ($executors) => $executors->whereKey($employee?->id ?? 0));

            if ($user->hasRole('organization-admin') && $employee) {
                $scope->orWhere('issues.organization_id', $employee->organization_id);
            }

            if ($user->hasRole('department-manager') && $employee?->department_id) {
                $scope->orWhere('issues.department_id', $employee->department_id)
                    ->orWhereHas('executors', fn ($executors) => $executors->where('employees.department_id', $employee->department_id));
            }
        });
    }

    public function reporter(): BelongsTo
    {
        return $this->belongsTo(Employee::class, 'reporter_employee_id');
    }

    public function organization(): BelongsTo
    {
        return $this->belongsTo(Organization::class);
    }

    public function department(): BelongsTo
    {
        return $this->belongsTo(Department::class);
    }

    public function category(): BelongsTo
    {
        return $this->belongsTo(IssueCategory::class, 'issue_category_id');
    }

    /**
     * The people doing the work — one or several, like a task's assignees.
     */
    public function executors(): BelongsToMany
    {
        return $this->belongsToMany(Employee::class, 'issue_executors')->withTimestamps();
    }

    public function resolvedBy(): BelongsTo
    {
        return $this->belongsTo(User::class, 'resolved_by');
    }

    public function comments(): HasMany
    {
        return $this->hasMany(IssueComment::class);
    }

    public function activities(): HasMany
    {
        return $this->hasMany(IssueActivity::class)->latest();
    }

    /**
     * Get the attributes that should be cast.
     *
     * @return array<string, string>
     */
    protected function casts(): array
    {
        return [
            'latitude' => 'decimal:7',
            'longitude' => 'decimal:7',
            'status' => IssueStatus::class,
            'resolved_at' => 'datetime',
        ];
    }
}
